agent/src/beerocks/fronthaul_manager/fronthaul_manager_main.cpp

```diff
@@ -147,60 +178,82 @@ int main(int argc, char *argv[])
+    // Kill running fronthaul and write pid file
+    std::string base_fronthaul_name = std::string(BEEROCKS_FRONTHAUL) + "_" + fronthaul_iface;
+    beerocks::os_utils::kill_pid(beerocks_slave_conf.temp_path, base_fronthaul_name);
+    beerocks::os_utils::write_pid_file(beerocks_slave_conf.temp_path, base_fronthaul_name);
     std::string pid_file_path =
-        beerocks_slave_conf.temp_path + "pid/" + base_monitor_name; // for file touching
-
-    // start monitor
-    std::string slave_uds =
-        beerocks_slave_conf.temp_path + std::string(BEEROCKS_SLAVE_UDS) + "_" + monitor_iface;
-    son::monitor_thread monitor(slave_uds, monitor_iface, beerocks_slave_conf, logger);
-    if (monitor.init()) {
-        auto touch_time_stamp_timeout = std::chrono::steady_clock::now();
-        while (g_running) {
-
-            // Handle signals
-            if (s_signal) {
-                handle_signal();
-                continue;
-            }
+        beerocks_slave_conf.temp_path + "pid/" + base_fronthaul_name; // For file touching
 
-            if (std::chrono::steady_clock::now() > touch_time_stamp_timeout) {
-                beerocks::os_utils::touch_pid_file(pid_file_path);
-                touch_time_stamp_timeout =
-                    std::chrono::steady_clock::now() +
-                    std::chrono::seconds(beerocks::TOUCH_PID_TIMEOUT_SECONDS);
-            }
+    // Get Agent UDS file
+    std::string agent_uds =
+        beerocks_slave_conf.temp_path + std::string(BEEROCKS_SLAVE_UDS) + "_" + fronthaul_iface;
+
+    // Create ap_manager
+    son::ap_manager_thread ap_manager(agent_uds, fronthaul_iface, *g_logger_ap_mananger);
+
+    if (!ap_manager.init()) {
+        CLOG(ERROR, g_logger_ap_mananger->get_logger_id()) << "ap manager init() has failed!";
+        return 0;
+    }
+
+    // Create Monitor
+    son::monitor_thread monitor(agent_uds, fronthaul_iface, beerocks_slave_conf, *g_logger_monitor);
 
-            if (!monitor.work()) {
+    auto touch_time_stamp_timeout = std::chrono::steady_clock::now();
+    while (g_running) {
+
+        // Handle signals
+        if (s_signal) {
+            handle_signal();
+            continue;
+        }
+
+        if (std::chrono::steady_clock::now() > touch_time_stamp_timeout) {
+            beerocks::os_utils::touch_pid_file(pid_file_path);
+            touch_time_stamp_timeout = std::chrono::steady_clock::now() +
+                                       std::chrono::seconds(beerocks::TOUCH_PID_TIMEOUT_SECONDS);
+        }
+
+        if (!ap_manager.work()) {
+            break;
+        }
+
+        // The ap_manager is the main process thread. After the ap_manager is finished the attach
+        // process, start the monitor thread. There is no point to start it before.
+        auto ap_manager_state = ap_manager.get_state();
+        if (ap_manager_state < son::ap_manager_thread::eApManagerState::ATTACHED) {
```

Comment:
This condition is not really needed.
It's enough to if the AP Manager state is `>= ATTACHED` and the monitor is not running...

Reply:
Yes, you right. Remove it and leave only the condition on the "else".